src/rbac/defaultRoleManager.ts

```diff
@@ -283,6 +283,10 @@ export class DefaultRoleManager implements RoleManager {
     return role1.hasRole(name2, this.maxHierarchyLevel);
   }
 
+  public async hasLink(name1: string, name2: string, ...domain: string[]): Promise<boolean> {
```

Comment:
Had to think this again. Does this actually do anything? Like none of the functions within syncedHasLink actually `await` for anything or handle promises at all. `generateTempRoles` is sync, so is `hasRole`. If they would return a promise, they would not work correctly anyway. 

Do we have test where underlying relations are promises? Does `enforcer` deal with this in some way? @nodece

Reply:
I am more or less confused whether defaultRoleManager can actually support promises. If not, we should just allow `RoleManager`interface to return sync results and then it's up to other `RoleManager` class definers to create functions that support them.

Reply:
Good catch! For built-in RoleManager, it without the async is better, but we must consider providing a set of async methods, maybe some users make their own RoleManager in casbin, and needs async methods, so we need to providing that.

Reply:
I do agree that async methods should be supported, but do prefer `DefaultRoleManager` to be either truly async or just sync.

Sync seems a better option in this case. However how to solve this? Easiest way would be to convert all functions back to sync and allow async implementations within RoleManager interface. However this is a breaking change.

I wouldn't matter too much if people are using await in async functions, but if they are using `hasLink.then()` it will throw an error.

Reply:
> I do agree that async methods should be supported, but do prefer DefaultRoleManager to be either truly async or just sync.

I agreed with you, but the sync method only applies to `enforceSync`, or the sync method is preferred.


> Sync seems a better option in this case. However how to solve this? Easiest way would be to convert all functions back to sync and allow async implementations within RoleManager interface. However this is a breaking change.

You can add a set of synced method like `syncedClear`, `syncedAddLink`, `syncedDeleteLink`, `syncedHasLink`, `syncedGetRoles`, `syncedGetUsers`, `syncedPrintRoles`.

Reply:
Yeah, that's an option. The only thing I do not like is that these function names can cause confusion with things like `syncedEnforcer`.

Reply:
@nodece What is the culprit function that causes async to be such a performance deficit? I think we can just ignore the synchronous issue if we can deal with async being closely representative of the sync function,

Reply:
> What is the culprit function that causes async to be such a performance deficit?

What means?

> I think we can just ignore the synchronous issue if we can make async closely representative of the sync function,

Do you mean that return `Promise<Value>` or `Value`?

Reply:
What I mean is that we could leave implementation fully async if enforcing permissions would not take forever.  Is it `compileAsync` which takes forever or the `privateEnforce` loop itself?

Reply:
> What I mean is that we could leave implementation fully async if enforcing permissions would not take forever

Yes, we still need to provide both async and sync methods.

 > Is it compileAsync which takes forever or the privateEnforce loop itself?

compileAsync/compile is required, if users want to use the async method in the matcher of model, we have to compileAsync, this is determined by the `expression-eval`.

Reply:
> compileAsync/compile is required, if users want to use the async method in the matcher of model, we have to compileAsync, this is determined by the `expression-eval`.

I am not asking the use case or reason why something is used, but why something is performing so poorly.

For example. If we are doing a promise loop, it probably isn't wise to await/yield each promise separately, but put all promises into an array and either use `Promise.all()`, `Promise.any()` or `Promise.race()` based on the use case.

If the performance is tanked by `compileAsync` itself, it might be a good idea to look if there is something to be done. There seems to be an alternative fork of the library since the original hasn't been updated or maintained for a year. https://github.com/donmccurdy/expression-eval/pull/63/files

I have not looked too deeply into which parts of the code caused the performance drop.

One option would also be to try to evaluate whether the function is async or whether it includes a promise. Based on that a different executor is used.

Reply:
> I am not asking the use case or reason why something is used, but why something is performing so poorly.

I'm not sure if it is affected by`generatorRunAsync` or `expression-eval`, we need time to optimize this.